File: utils/crypto.ts
import { ALPHABET, HEBREW_ALPHABET, GEMATRIA_VALUES, HIEROGLYPH_UNILITERALS } from '../constants.ts';
import { GematriaSchema } from '../types.ts';

// --- Core Decryption Logic ---

export const vigenereDecrypt = (ciphertext: string, key: string): string => {
    if (!key) return ciphertext;
    const sanitizedKey = key.toUpperCase().replace(/[^A-Z]/g, '');
    if (!sanitizedKey) return ciphertext;
    
    let decrypted = '';
    let keyIndex = 0;
    const upperCiphertext = ciphertext.toUpperCase();

    for (let i = 0; i < upperCiphertext.length; i++) {
        const char = upperCiphertext[i];
        if (ALPHABET.includes(char)) {
            const keyChar = sanitizedKey[keyIndex % sanitizedKey.length];
            const shift = ALPHABET.indexOf(keyChar);
            const decryptedCharIndex = (ALPHABET.indexOf(char) - shift + 26) % 26;
            decrypted += ALPHABET[decryptedCharIndex];
            keyIndex++;
        } else {
            // Keep non-alphabetic characters as they are, but don't advance key
            decrypted += char;
        }
    }
    return decrypted;
};

export const caesarDecrypt = (ciphertext: string, shift: number): string => {
    const upperCiphertext = ciphertext.toUpperCase();
    let decrypted = '';
    for (let i = 0; i < upperCiphertext.length; i++) {
        const char = upperCiphertext[i];
        if (ALPHABET.includes(char)) {
            const decryptedCharIndex = (ALPHABET.indexOf(char) - shift + 26) % 26;
            decrypted += ALPHABET[decryptedCharIndex];
        } else {
            decrypted += char;
        }
    }
    return decrypted;
};

export const atbashDecrypt = (ciphertext: string): string => {
    const reversedAlphabet = ALPHABET.split('').reverse().join('');
    const reversedHebrew = HEBREW_ALPHABET.split('').reverse().join('');
    
    let decrypted = '';
    for (const char of ciphertext) {
        let index = ALPHABET.indexOf(char.toUpperCase());
        if (index !== -1) {
            decrypted += reversedAlphabet[index];
            continue;
        }
        index = HEBREW_ALPHABET.indexOf(char);
        if (index !== -1) {
            decrypted += reversedHebrew[index];
            continue;
        }
        decrypted += char;
    }
    return decrypted;
};


// --- Statistical Analysis ---

export const calculateFrequencies = (text: string): { letter: string; frequency: number }[] => {
    const counts: { [key: string]: number } = {};
    const upperText = text.toUpperCase().replace(/[^A-Z]/g, '');
    const textLength = upperText.length;

    if (textLength === 0) {
        return ALPHABET.split('').map(letter => ({ letter, frequency: 0 }));
    }

    for (const letter of ALPHABET) {
        counts[letter] = 0;
    }

    for (const char of upperText) {
        if (counts[char] !== undefined) {
            counts[char]++;
        }
    }
    
    return ALPHABET.split('').map(letter => ({
        letter,
        frequency: (counts[letter] || 0) / textLength,
    }));
};

export const calculateIoC = (text: string): number => {
    const counts: { [key: string]: number } = {};
    const upperText = text.toUpperCase().replace(/[^A-Z]/g, '');
    const N = upperText.length;

    if (N < 2) return 0;

    for (const char of upperText) {
        counts[char] = (counts[char] || 0) + 1;
    }

    let sum = 0;
    for (const letter of ALPHABET) {
        const ni = counts[letter] || 0;
        sum += ni * (ni - 1);
    }

    return sum / (N * (N - 1));
};

export const calculateNgrams = (text: string, n: number): Map<string, number> => {
    const ngrams = new Map<string, number>();
    // Do not convert to upper for esoteric texts
    const cleanedText = text.replace(/[\s.,?।“”]/g, '');
    if (cleanedText.length < n) {
        return ngrams;
    }

    for (let i = 0; i <= cleanedText.length - n; i++) {
        const ngram = cleanedText.substring(i, i + n);
        ngrams.set(ngram, (ngrams.get(ngram) || 0) + 1);
    }
    return ngrams;
};

// --- Kasiski Examination ---

const getFactors = (num: number): number[] => {
    const factors = new Set<number>();
    for (let i = 1; i <= Math.sqrt(num); i++) {
        if (num % i === 0) {
            factors.add(i);
            factors.add(num / i);
        }
    }
    return Array.from(factors).sort((a, b) => a - b);
};

export interface SequenceInfo {
    sequence: string;
    positions: number[];
    distances: number[];
}

export const findRepeatedSequences = (text: string, minLength = 3, maxLength = 6): Map<string, number[]> => {
    const sequences = new Map<string, number[]>();
    const upperText = text.toUpperCase().replace(/[^A-Z]/g, '');
    const len = upperText.length;

    for (let l = maxLength; l >= minLength; l--) {
        for (let i = 0; i <= len - l; i++) {
            const seq = upperText.substring(i, i + l);
            if (!sequences.has(seq)) {
                sequences.set(seq, [i]);
            } else {
                sequences.get(seq)!.push(i);
            }
        }
    }

    const repeatedSequences = new Map<string, number[]>();
    for (const [seq, positions] of sequences.entries()) {
        if (positions.length > 1) {
            repeatedSequences.set(seq, positions);
        }
    }
    return repeatedSequences;
};

export const analyzeDistances = (sequences: Map<string, number[]>): { factors: Map<number, number>; sequenceDetails: SequenceInfo[] } => {
    const factorCounts = new Map<number, number>();
    const sequenceDetails: SequenceInfo[] = [];

    for (const [sequence, positions] of sequences.entries()) {
        const distances: number[] = [];
        for (let i = 1; i < positions.length; i++) {
            const distance = positions[i] - positions[0];
            distances.push(distance);
            const factors = getFactors(distance);
            for (const factor of factors) {
                if(factor > 1 && factor <= 20) { 
                    factorCounts.set(factor, (factorCounts.get(factor) || 0) + 1);
                }
            }
        }
        sequenceDetails.push({ sequence, positions, distances });
    }

    return { factors: factorCounts, sequenceDetails };
};

export const calculateIoCForColumns = (text: string, keyLength: number): number => {
    if (keyLength <= 0) return 0;
    const upperText = text.toUpperCase().replace(/[^A-Z]/g, '');
    if (upperText.length < keyLength) return 0;
    
    const columns: string[] = Array(keyLength).fill('');

    for (let i = 0; i < upperText.length; i++) {
        columns[i % keyLength] += upperText[i];
    }
    
    const iocSum = columns.reduce((sum, col) => sum + calculateIoC(col), 0);
    return iocSum / keyLength;
};

// --- Textual & Esoteric Analysis ---

export const getUniqueChars = (text: string): string[] => {
    return Array.from(new Set(text.split(''))).sort();
};


export const calculateGematria = (text: string, schema: GematriaSchema): number => {
    const values = GEMATRIA_VALUES[schema];
    let sum = 0;
    
    if (schema === 'sanskrit_katapayadi') {
        // Normalize to separate base characters from combining marks (vowel signs), then remove the marks.
        const normalizedText = text.normalize('NFD').replace(/[\u093e-\u094c\u094d\u0902\u0903]/g, '');
        for (const char of normalizedText) {
            if (values[char] !== undefined) {
                sum += values[char];
            }
        }
        return sum;
    }
    
    const textToProcess = (schema === 'pythagorean' || schema === 'chaldean' || schema === 'latin_roman') ? text.toUpperCase() : text;
    for (const char of textToProcess) {
        if (values[char] !== undefined) {
            sum += values[char];
        }
    }
    return sum;
};

export const transliterateToHieroglyphs = (text: string): string => {
    if (!text) return '';
    const upperText = text.toUpperCase();
    let result = '';
    let i = 0;
    while (i < upperText.length) {
        // Check for two-character glyphs first (e.g., 'SH', 'TH')
        if (i + 1 < upperText.length) {
            const twoChar = upperText.substring(i, i + 2);
            if (HIEROGLYPH_UNILITERALS[twoChar]) {
                result += HIEROGLYPH_UNILITERALS[twoChar].glyph;
                i += 2;
                continue;
            }
        }
        // Check for one-character glyphs
        const oneChar = upperText[i];
        if (HIEROGLYPH_UNILITERALS[oneChar]) {
            result += HIEROGLYPH_UNILITERALS[oneChar].glyph;
        }
        i++;
    }
    return result;
};

File: app/services/crypto.py
import math
import unicodedata
import re
from app.services.constants import (
    ALPHABET, HEBREW_ALPHABET, GEMATRIA_VALUES, HIEROGLYPH_UNILITERALS,
)


def vigenere_decrypt(ciphertext: str, key: str) -> str:
    if not key:
        return ciphertext
    sanitized_key = re.sub(r'[^A-Z]', '', key.upper())
    if not sanitized_key:
        return ciphertext

    decrypted: list[str] = []
    key_index = 0
    upper_ciphertext = ciphertext.upper()
    key_len = len(sanitized_key)

    for char in upper_ciphertext:
        if char in ALPHABET:
            shift = ALPHABET.index(sanitized_key[key_index % key_len])
            decrypted_index = (ALPHABET.index(char) - shift) % 26
            decrypted.append(ALPHABET[decrypted_index])
            key_index += 1
        else:
            decrypted.append(char)

    return ''.join(decrypted)


def caesar_decrypt(ciphertext: str, shift: int) -> str:
    shift = shift % 26
    decrypted: list[str] = []
    for char in ciphertext.upper():
        if char in ALPHABET:
            decrypted_index = (ALPHABET.index(char) - shift) % 26
            decrypted.append(ALPHABET[decrypted_index])
        else:
            decrypted.append(char)
    return ''.join(decrypted)


def atbash_decrypt(ciphertext: str) -> str:
    reversed_alphabet = ALPHABET[::-1]
    reversed_hebrew = HEBREW_ALPHABET[::-1]
    decrypted: list[str] = []
    for char in ciphertext:
        idx = ALPHABET.find(char.upper())
        if idx != -1:
            decrypted.append(reversed_alphabet[idx])
            continue
        idx = HEBREW_ALPHABET.find(char)
        if idx != -1:
            decrypted.append(reversed_hebrew[idx])
            continue
        decrypted.append(char)
    return ''.join(decrypted)


def calculate_frequencies(text: str) -> list[dict]:
    counts: dict[str, int] = {}
    upper_text = re.sub(r'[^A-Z]', '', text.upper())
    text_length = len(upper_text)

    if text_length == 0:
        return [{'letter': letter, 'frequency': 0.0} for letter in ALPHABET]

    for letter in ALPHABET:
        counts[letter] = 0

    for char in upper_text:
        counts[char] = counts.get(char, 0) + 1

    return [
        {'letter': letter, 'frequency': counts[letter] / text_length}
        for letter in ALPHABET
    ]


def calculate_ioc(text: str) -> float:
    upper_text = re.sub(r'[^A-Z]', '', text.upper())
    N = len(upper_text)
    if N < 2:
        return 0.0

    counts: dict[str, int] = {}
    for char in upper_text:
        counts[char] = counts.get(char, 0) + 1

    total = sum(ni * (ni - 1) for ni in counts.values())
    return total / (N * (N - 1))


def calculate_ngrams(text: str, n: int) -> dict[str, int]:
    ngrams: dict[str, int] = {}
    cleaned = re.sub(r'[\s.,?\u201c\u201d।]', '', text)
    if len(cleaned) < n:
        return ngrams

    for i in range(len(cleaned) - n + 1):
        ngram = cleaned[i:i + n]
        ngrams[ngram] = ngrams.get(ngram, 0) + 1

    return ngrams


def _get_factors(num: int) -> list[int]:
    if num <= 0:
        return []
    factors: set[int] = set()
    for i in range(1, int(math.isqrt(num)) + 1):
        if num % i == 0:
            factors.add(i)
            factors.add(num // i)
    return sorted(factors)


def find_repeated_sequences(
    text: str, min_length: int = 3, max_length: int = 6
) -> dict[str, list[int]]:
    sequences: dict[str, list[int]] = {}
    upper_text = re.sub(r'[^A-Z]', '', text.upper())
    text_len = len(upper_text)

    for l in range(max_length, min_length - 1, -1):
        for i in range(text_len - l + 1):
            seq = upper_text[i:i + l]
            if seq not in sequences:
                sequences[seq] = [i]
            else:
                sequences[seq].append(i)

    return {
        seq: positions
        for seq, positions in sequences.items()
        if len(positions) > 1
    }


def analyze_distances(
    sequences: dict[str, list[int]]
) -> tuple[dict[int, int], list[dict]]:
    factor_counts: dict[int, int] = {}
    sequence_details: list[dict] = []

    for sequence, positions in sequences.items():
        distances: list[int] = []
        for i in range(1, len(positions)):
            distance = positions[i] - positions[0]
            distances.append(distance)
            for factor in _get_factors(distance):
                if 1 < factor <= 20:
                    factor_counts[factor] = factor_counts.get(factor, 0) + 1
        sequence_details.append({
            'sequence': sequence,
            'positions': positions,
            'distances': distances,
        })

    return factor_counts, sequence_details


def calculate_ioc_for_columns(text: str, key_length: int) -> float:
    if key_length <= 0:
        return 0.0
    upper_text = re.sub(r'[^A-Z]', '', text.upper())
    if len(upper_text) < key_length:
        return 0.0

    columns: list[str] = [''] * key_length
    for i, char in enumerate(upper_text):
        columns[i % key_length] += char

    ioc_sum = sum(calculate_ioc(col) for col in columns)
    return ioc_sum / key_length


def get_unique_chars(text: str) -> list[str]:
    return sorted(set(text))


def calculate_gematria(text: str, schema: str) -> int:
    values = GEMATRIA_VALUES[schema]
    total = 0

    if schema == 'sanskrit_katapayadi':
        normalized = unicodedata.normalize('NFD', text)
        cleaned = re.sub(r'[\u093e-\u094c\u094d\u0902\u0903]', '', normalized)
        for char in cleaned:
            total += values.get(char, 0)
        return total

    if schema in ('pythagorean', 'chaldean', 'latin_roman'):
        text = text.upper()

    for char in text:
        total += values.get(char, 0)

    return total


def transliterate_to_hieroglyphs(text: str) -> str:
    if not text:
        return ''
    upper_text = text.upper()
    result: list[str] = []
    i = 0
    while i < len(upper_text):
        if i + 1 < len(upper_text):
            two_char = upper_text[i:i + 2]
            if two_char in HIEROGLYPH_UNILITERALS:
                result.append(HIEROGLYPH_UNILITERALS[two_char]['glyph'])
                i += 2
                continue
        one_char = upper_text[i]
        if one_char in HIEROGLYPH_UNILITERALS:
            result.append(HIEROGLYPH_UNILITERALS[one_char]['glyph'])
        i += 1
    return ''.join(result)
